packages/Chem/src/chem-searches.ts

```diff
@@ -128,16 +129,17 @@ function saveFingerprintsToCol(col: DG.Column, fgs: Uint8Array[], fingerprintsTy
 }
 
 async function getUint8ArrayFingerprints(
-  molCol: DG.Column, fingerprintsType: Fingerprint = Fingerprint.Morgan, useSection = true): Promise<Uint8Array[]> {
+  molCol: DG.Column, fingerprintsType: Fingerprint = Fingerprint.Morgan, useSection = true, createMols = true): Promise<Uint8Array[]> {
   if (useSection)
     await chemBeginCriticalSection();
   try {
     const fgsCheck = checkForFingerprintsColumn(molCol, fingerprintsType);
     if (fgsCheck)
       return fgsCheck;
     else {
-      await _invalidate(molCol);
-      const fingerprints = await (await getRdKitService()).getFingerprints(fingerprintsType);
+      await _invalidate(molCol, createMols);
+      const molecules = createMols ? undefined : molCol.toList();
+      const fingerprints = await (await getRdKitService()).getFingerprints(fingerprintsType, molecules);
       saveFingerprintsToCol(molCol, fingerprints, fingerprintsType);
       chemEndCriticalSection();
```

Comment:
Do we understand how it all works? What does "useSection" do? Is it really ok to call "chemEndCriticalSection" (whatever this is) if useSection == false?

Reply:
Yes, it is ok, because as I see chemEndCriticalSection (which on its side calls criticalSectionEnd() fucntion) performs additional check for existance of exact key in global dictionary. And in case we didn't enter chemBeginCriticalSection() before caling chemEndCriticalSection we will not fail. But anyway we need to explore this part of code more.